packages/shared/src/hooks/useCompanionTrigger.tsx

```diff
@@ -0,0 +1,167 @@
+    row?: number,
+    column?: number,
+  ) => Promise<void>,
+): CompanionTrigger {
+  const isExtension = process.env.TARGET_BROWSER;
+  const { trackEvent } = useAnalyticsContext();
+  const { closeModal: closeLazyModal, openModal: openLazyModal } =
+    useLazyModal();
+  const { isActionsFetched, checkHasCompleted, completeAction } = useActions();
+  const gb = useGrowthBook();
+
+  const featureEnabled = useMemo(() => {
+    // gb.isOn triggers enrollment in the experiment
+    // to avoid noise in the data, only call this if we are in the extension
+    if (isExtension)
+      return gb.isOn(Features.EngagementLoopJuly2023Companion.id);
+
+    return false;
+  }, [gb, isExtension]);
+
+  const { requestContentScripts, useContentScriptStatus } =
+    useExtensionPermission({
+      origin: 'companion modal permission button',
+    });
+
+  const { contentScriptGranted } = useContentScriptStatus();
+  const alreadyCompleted = checkHasCompleted(
+    ActionType.EngagementLoopJuly2023CompanionModal,
+  );
+
+  const handleCompleteAction = useCallback(() => {
+    if (isActionsFetched)
+      completeAction(ActionType.EngagementLoopJuly2023CompanionModal);
+  }, [isActionsFetched, completeAction]);
+
+  const closeModal = useCallback(() => {
+    closeLazyModal();
+    handleCompleteAction();
+  }, [closeLazyModal, handleCompleteAction]);
+
+  const activateCompanion = useCallback(async () => {
+    await requestContentScripts();
+    closeModal();
+    handleCompleteAction();
+  }, [requestContentScripts, closeModal, handleCompleteAction]);
+
+  const openArticle: OpenArticle = useCallback(
+    (data) => async () => {
+      await customPostClickHandler(
+        data.post,
+        data.index,
+        data.row,
+        data.column,
+      );
+
+      closeModal();
+      handleCompleteAction();
+    },
+    [customPostClickHandler, closeModal, handleCompleteAction],
+  );
+
+  const openModal = useCallback(
+    (data?: CompanionTriggerProps) => {
+      openLazyModal({
+        type: LazyModal.CompanionModal,
+        props: {
+          url: data?.post?.permalink,
+          onReadArticleClick: openArticle(data),
+          onActivateCompanion: activateCompanion,
+          onAfterClose() {
+            trackEvent({
+              event_name: AnalyticsEvent.CloseCompanionPermissionModal,
+              extra: JSON.stringify({ origin: Origin.Auto }),
+            });
+          },
+        },
+      });
+
+      trackEvent({
+        event_name: AnalyticsEvent.OpenCompanionPermissionModal,
+        extra: JSON.stringify({ origin: Origin.Auto }),
+      });
+    },
+    [openArticle, activateCompanion, openLazyModal, trackEvent],
+  );
+
+  const articleClickHandler = useCallback(
+    async (
+      e: React.MouseEvent,
+      { post, index, row, column }: Partial<CompanionTriggerProps> = {},
+    ) => {
+      if (
+        !isExtension ||
+        !featureEnabled ||
+        alreadyCompleted ||
+        contentScriptGranted
```

Comment:
You can pull this condition outside of `useCallback` and call it something meaningful. Eg. `shouldOverwritePostClickHandler`. It will make it easier to someone to understand what it checks and also lower the amount dependencies for `useCallback`.